logserver/src/main/java/alluxio/logserver/AlluxioLogServerProcess.java

```diff
@@ -0,0 +1,180 @@
+    SynchronousQueue<Runnable> synchronousQueue = new SynchronousQueue<>();
+    mThreadPool =
+        new ThreadPoolExecutor(mMinNumberOfThreads, mMaxNumberOfThreads,
+            THREAD_KEEP_ALIVE_TIME_MS, TimeUnit.MILLISECONDS, synchronousQueue);
+    mStopped = false;
+    try {
+      mServerSocket = new ServerSocket(mPort);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to bind to port {}.", e);
+    }
+    while (!mStopped) {
+      Socket client;
+      try {
+        client = mServerSocket.accept();
+      } catch (IOException e) {
+        try {
+          ServerSocket tmpServerSocket = new ServerSocket(mPort);
+          // If the code reaches here, it indicates mServerSocket has been closed.
+          // Therefore, by assigning the reference, we have successfully re-binded
+          // mServerSocket.
+          mServerSocket = tmpServerSocket;
+        } catch (IOException e1) {
+          // The original mServerSocket is not closed, therefore just ignore the exception
+        }
+        continue;
+      }
+      InetAddress inetAddress = client.getInetAddress();
+      AlluxioLog4jSocketNode clientSocketNode = new AlluxioLog4jSocketNode(this, client);
+      while (true) {
+        try {
+          mThreadPool.execute(clientSocketNode);
+          break;
+        } catch (RejectedExecutionException e) {
+          // Alluxio log clients (master, secondary master, proxy and workers establish
+          // long-living connections with the log server. Therefore, if the log server cannot
+          // find a thread to service a log client, it is very likely due to low number of
+          // worker threads. If retry fails, then it makes sense just to let system throw
+          // an exception. The system admin should increase the thread pool size.
+          throw new RuntimeException(
+              "Increase the number of worker threads in the thread pool", e);
+        } catch (Error | Exception e) {
+          LOG.error("ExecutorService threw error: ", e);
+          throw e;
+        }
+      }
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * Stop the main thread of {@link AlluxioLogServerProcess}.
+   *
+   * Close the server socket, shutdown the thread pool, stop accepting new requests,
+   * and blocks until all worker threads terminate.
+   */
+  @Override
+  public void stop() throws Exception {
+    mStopped = true;
+    if (mServerSocket != null) {
+      try {
+        mServerSocket.close();
+      } catch (IOException e) {
+        LOG.warn("Exception in closing server socket.", e);
+      }
+    }
+
+    mThreadPool.shutdown();
+    try {
+      boolean ret = mThreadPool.awaitTermination(THREAD_KEEP_ALIVE_TIME_MS, TimeUnit.MILLISECONDS);
+      if (ret) {
+        LOG.info("All worker threads have terminated.");
+      } else {
+        LOG.warn("Log server has timedout waiting for worker threads to terminate.");
+      }
+    } catch (InterruptedException e) {
```

Comment:
since this method throws `Exception`, the caller needs to deal with `InterruptedException` anyway, so we can allow it to propagate instead of wrapping it

Reply:
Done.